File: src/main/java/org/kite9/diagram/common/algorithms/det/UnorderedSet.java
package org.kite9.diagram.common.algorithms.det;

import java.util.Collection;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Objects;
import java.util.Spliterator;
import java.util.function.Consumer;
import java.util.stream.Stream;

/**
 * This prevents any attempts to order the contents, as the order will always 
 * be non-deterministic if we use objects that don't declare hashCode.  
 * 
 * Use this in preference to HashSet, unless ordering is required, and in which case use
 * LinkedHashSet or DetHashSet.
 * 
 * @author robmoffat
 *
 */
public class UnorderedSet<K> extends HashSet<K>{

	public UnorderedSet() {
		super();
	}

	public UnorderedSet(Collection<? extends K> c) {
		super(c);
	}

	public UnorderedSet(int initialCapacity, float loadFactor) {
		super(initialCapacity, loadFactor);
	}

	public UnorderedSet(int initialCapacity) {
		super(initialCapacity);
	}

	@Override
	public Stream<K> stream() {
		throw new UnsupportedOperationException("Prevented By UnorderedSet");
	}

	@Override
	public Stream<K> parallelStream() {
		throw new UnsupportedOperationException("Prevented By UnorderedSet");
	}

	@Override
	public void forEach(Consumer<? super K> action) {
		throw new UnsupportedOperationException("Prevented By UnorderedSet");
	}

	@Override
	public Iterator<K> iterator() {
		throw new UnsupportedOperationException("Prevented By UnorderedSet");
	}

	@Override
	public Spliterator<K> spliterator() {
		throw new UnsupportedOperationException("Prevented By UnorderedSet");
	}

	/**
	 * Use sparingly - provided for logging only.
	 */
	@Override
	public Object[] toArray() {
		Object[] r = new Object[size()];
        int i = 0;
        for (Iterator<K> iterator = super.iterator(); iterator.hasNext();) {
			K type = iterator.next();
			r[i] = type;
			i++;
		}
        
        return r;
	}
	
	 public boolean retainAll(Collection<?> c) {
	        Objects.requireNonNull(c);
	        boolean modified = false;
	        Iterator<K> it = super.iterator();
	        while (it.hasNext()) {
	            if (!c.contains(it.next())) {
	                it.remove();
	                modified = true;
	            }
	        }
	        return modified;
	    }
	
}

File: src/main/kotlin/org/kite9/diagram/common/algorithms/det/UnorderedSet.kt
package org.kite9.diagram.common.algorithms.det


/**
 * This prevents any attempts to order the contents, as the order will always
 * be non-deterministic if we use objects that don't declare hashCode.
 *
 * Use this in preference to HashSet, unless ordering is required, and in which case use
 * LinkedHashSet or DetHashSet.
 *
 * @author robmoffat
 */
class UnorderedSet<K> : HashSet<K> {

    constructor() : super() {}
    constructor(c: Collection<K>) : super(c) {}
    constructor(initialCapacity: Int) : super(initialCapacity) {}


    /**
     * Use sparingly - provided for logging only.
     */
    override fun toArray(): Array<Any?> {
        val r = arrayOfNulls<Any>(size)
        var i = 0
        val iterator: Iterator<K> = super.iterator()
        while (iterator.hasNext()) {
            val type = iterator.next()
            r[i] = type
            i++
        }
        return r
    }

    override fun retainAll(elements: Collection<K>): Boolean {
        var modified = false
        val it = super.iterator()
        while (it.hasNext()) {
            if (!elements.contains(it.next())) {
                it.remove()
                modified = true
            }
        }
        return modified
    }
}